Add riscv64 to supported architectures

Tested to work fine on FreeBSD. I'm not sending a PR, because FreeBSD is not in the table, but I suspect it should work fine on Linux as well.

@pkubaj can you please give #81 a try on RiscV64 also? 

Compilation with GCC failed on Linux because of linker errors (undefined reference to __atomic_compare_exchange_1 and __atomic_exchange_1) even with -latomic. But that seems to be a GCC problem, since you were successful with clang.

--- a/README.md
+++ b/README.md
@@ -76,8 +76,7 @@ The Fraunhofer Versatile Video Decoder (VVdeC) is a fast H.266/VVC software deco
 
 :white_check_mark: tested and works :black_square_button: needs testing (might already work) :x: does not work
 
-There is anecdotal evidence that VVdeC also runs on other architectures and platforms not listed here. For more information see the
-[Wiki](https://github.com/fraunhoferhhi/vvdec/wiki#supported-architectures).
+Other architectures and platforms might work, see the [Wiki](https://github.com/fraunhoferhhi/vvdec/wiki#supported-architectures).
 
 ## Information
 

--- a/README.md
+++ b/README.md
@@ -74,7 +74,10 @@ The Fraunhofer Versatile Video Decoder (VVdeC) is a fast H.266/VVC software deco
   </tr>
 </table>
 
-:white_check_mark: tested and works :black_square_button: needs testing (might aleady work) :x: does not work
+:white_check_mark: tested and works :black_square_button: needs testing (might already work) :x: does not work
+
+There is anecdotal evidence that VVdeC also runs on other architectures and platforms not listed here. For more information see the
+[Wiki](https://github.com/fraunhoferhhi/vvdec/wiki#supported-architectures).
 
 ## Information
 
@@ -110,7 +113,7 @@ For multi-configuration projects (e.g. Visual Studio or Xcode) specify `--config
 
 ### How to build using GNU Make?
 
-On top of the CMake build system, convinence Makefile is provided to simplify the build process. To build using GNU Make please run the following:
+On top of the CMake build system, convenience Makefile is provided to simplify the build process. To build using GNU Make please run the following:
 
 ```sh
 make install-release <options>

--- a/include/vvdec/vvdec.h
+++ b/include/vvdec/vvdec.h
@@ -136,6 +136,9 @@ typedef enum
 #elif VVDEC_ARCH_WASM
   VVDEC_SIMD_WASM     = 2,
   VVDEC_SIMD_MAX      = VVDEC_SIMD_WASM
+#else
+  VVDEC_SIMD_SIMDE_ANY= 2,
+  VVDEC_SIMD_MAX      = VVDEC_SIMD_SIMDE_ANY
 #endif
 }vvdecSIMD_Extension;
 

--- a/source/App/vvdecapp/CmdLineParser.h
+++ b/source/App/vvdecapp/CmdLineParser.h
@@ -323,6 +323,8 @@ public:
           case VVDEC_SIMD_NEON:    cll = "NEON";      break;
 #elif VVDEC_ARCH_WASM
           case VVDEC_SIMD_WASM:    cll = "WASM-SIMD"; break;
+#else
+          case VVDEC_SIMD_SIMDE_ANY:cll = "SIMDE-ANY"; break;
 #endif
           default:                 return -1;
           };

--- a/source/Lib/CommonLib/UnitTools.cpp
+++ b/source/Lib/CommonLib/UnitTools.cpp
@@ -1241,32 +1241,32 @@ void PU::getInterMergeCandidates( const PredictionUnit &pu, MergeCtx& mrgCtx, Mo
     }
   }
 
-  uint32_t uiArrayAddr = cnt;
+  int iArrayAddr = cnt;
 
   int iNumRefIdx = slice.isInterB() ? std::min(slice.getNumRefIdx(REF_PIC_LIST_0), slice.getNumRefIdx(REF_PIC_LIST_1)) : slice.getNumRefIdx(REF_PIC_LIST_0);
 
   int r = 0;
   int refcnt = 0;
   // second condition needed for gcc-10 overflow checking. Required for now. TODO: fix properly
-  while (uiArrayAddr < maxNumMergeCand && uiArrayAddr < MRG_MAX_NUM_CANDS)
+  while (iArrayAddr < maxNumMergeCand && iArrayAddr < MRG_MAX_NUM_CANDS)
   {
-    mrgCtx.interDirNeighbours [uiArrayAddr     ] = 1;
-    mrgCtx.BcwIdx             [uiArrayAddr     ] = BCW_DEFAULT;
-    mrgCtx.mvFieldNeighbours  [uiArrayAddr << 1] . setMvField(Mv(0, 0), r);
-    mrgCtx.useAltHpelIf       [uiArrayAddr     ] = false;
+    mrgCtx.interDirNeighbours [iArrayAddr     ] = 1;
+    mrgCtx.BcwIdx             [iArrayAddr     ] = BCW_DEFAULT;
+    mrgCtx.mvFieldNeighbours  [iArrayAddr << 1] . setMvField(Mv(0, 0), r);
+    mrgCtx.useAltHpelIf       [iArrayAddr     ] = false;
 
     if (slice.isInterB())
     {
-      mrgCtx.interDirNeighbours [ uiArrayAddr          ] = 3;
-      mrgCtx.mvFieldNeighbours  [(uiArrayAddr << 1) + 1].setMvField(Mv(0, 0), r);
+      mrgCtx.interDirNeighbours [ iArrayAddr          ] = 3;
+      mrgCtx.mvFieldNeighbours  [(iArrayAddr << 1) + 1].setMvField(Mv(0, 0), r);
     }
 
-    if ( mrgCtx.interDirNeighbours[uiArrayAddr] == 1 && cu.slice->getRefPOC(REF_PIC_LIST_0, mrgCtx.mvFieldNeighbours[uiArrayAddr << 1].mfRefIdx) == cu.slice->getPOC())
+    if ( mrgCtx.interDirNeighbours[iArrayAddr] == 1 && cu.slice->getRefPOC(REF_PIC_LIST_0, mrgCtx.mvFieldNeighbours[iArrayAddr << 1].mfRefIdx) == cu.slice->getPOC())
     {
-      mrgCtx.mrgTypeNeighbours[uiArrayAddr] = MRG_TYPE_IBC;
+      mrgCtx.mrgTypeNeighbours[iArrayAddr] = MRG_TYPE_IBC;
     }
 
-    uiArrayAddr++;
+    iArrayAddr++;
 
     if (refcnt == iNumRefIdx - 1)
     {
@@ -1278,7 +1278,7 @@ void PU::getInterMergeCandidates( const PredictionUnit &pu, MergeCtx& mrgCtx, Mo
       ++refcnt;
     }
   }
-  mrgCtx.numValidMergeCand = uiArrayAddr;
+  mrgCtx.numValidMergeCand = iArrayAddr;
 }
 
 bool PU::checkDMVRCondition( const PredictionUnit& pu )

--- a/source/Lib/CommonLib/x86/CommonDefX86.cpp
+++ b/source/Lib/CommonLib/x86/CommonDefX86.cpp
@@ -281,6 +281,8 @@ std::string read_simd_extension_name()
     return std::string( "NEON/SIMDE(" ) + vext_names[vext] + ")" ;
 #  elif defined( REAL_TARGET_WASM )
     return std::string( "WASM/Emscripten(" ) + vext_names[vext] + ")";
+#  else
+    return std::string( "SIMDE(" ) + vext_names[vext] + ")" ;
 #  endif
   }
 # endif   // !REAL_TARGET_X86

--- a/source/Lib/CommonLib/x86/InterpolationFilterX86.h
+++ b/source/Lib/CommonLib/x86/InterpolationFilterX86.h
@@ -1487,6 +1487,11 @@ static void simdInterpolateN2_2D( const ClpRng& clpRng, const Pel* src, const pt
 #endif
     __m128i mmLast4H;
 
+#ifndef REAL_TARGET_X86
+    // on some platforms gcc thinks this is used uninitialized with simd-everywhere
+    mmLastH[0] = _mm_setzero_si128();
+#endif
+
     for( int row = -1; row < height; row++ )
     {
       _mm_prefetch( ( const char * ) &src[srcStride], _MM_HINT_T0 );

--- a/source/Lib/CommonLib/x86/IntraPredX86.h
+++ b/source/Lib/CommonLib/x86/IntraPredX86.h
@@ -452,6 +452,11 @@ void  IntraPredSampleFilter_SIMD(Pel *ptrSrc,const ptrdiff_t srcStride,PelBuf &p
     __m256i vbdmax   = _mm256_set1_epi32( clpRng.max() );
     __m256i wl16,wl16start;
 
+#ifndef REAL_TARGET_X86
+    // on some platforms gcc thinks this is used uninitialized with simd-everywhere
+    wl16start = _mm256_setzero_si256();
+#endif
+
     if (scale==0)
     {
       wl16start = _mm256_set_epi16(0,0,0,0,0,0,0,0,0,0,0,0,0,2,8,32);

--- a/source/Lib/libmd5/MD5.h
+++ b/source/Lib/libmd5/MD5.h
@@ -59,11 +59,11 @@ static const uint32_t MD5_DIGEST_STRING_LENGTH=16;
 #ifndef __BIG_ENDIAN__
 # define byteReverse(buf, len)    /* Nothing */
 #else
-void byteReverse(uint32_t *buf, unsigned len);
+static void byteReverse(uint32_t *buf, unsigned len);
 /*
  * Note: this code is harmless on little-endian machines.
  */
-void byteReverse(uint32_t *buf, unsigned len)
+static void byteReverse(uint32_t *buf, unsigned len)
 {
   uint32_t t;
   do {

--- a/thirdparty/simde/x86/sse2.h
+++ b/thirdparty/simde/x86/sse2.h
@@ -5893,9 +5893,9 @@ simde_mm_slli_epi16 (simde__m128i a, const int imm8)
 //        ((imm8) > 15) ? \
 //          vandq_s16(simde__m128i_to_neon_i16(a), vdupq_n_s16(0)) : \
 //          vshlq_n_s16(simde__m128i_to_neon_i16(a), ((imm8) & 15))))
-//#elif defined(SIMDE_WASM_SIMD128_NATIVE)
-//  #define simde_mm_slli_epi16(a, imm8) \
-//    ((imm8 < 16) ? wasm_i16x8_shl(simde__m128i_to_private(a).wasm_v128, imm8) : wasm_i16x8_const(0,0,0,0,0,0,0,0))
+#elif defined(SIMDE_WASM_SIMD128_NATIVE)
+  #define simde_mm_slli_epi16(a, imm8) \
+    ((imm8 < 16) ? wasm_i16x8_shl(simde__m128i_to_private(a).wasm_v128, imm8) : wasm_i16x8_const(0,0,0,0,0,0,0,0))
 //#elif defined(SIMDE_POWER_ALTIVEC_P6_NATIVE)
 //  #define simde_mm_slli_epi16(a, imm8) \
 //    ((imm8 & ~15) ? simde_mm_setzero_si128() : simde__m128i_from_altivec_i16(vec_sl(simde__m128i_to_altivec_i16(a), vec_splat_u16(HEDLEY_STATIC_CAST(unsigned short, imm8)))))
@@ -5939,24 +5939,24 @@ simde_mm_slli_epi32 (simde__m128i a, const int imm8)
 //        ((imm8) > 31) ? \
 //          vandq_s32(simde__m128i_to_neon_i32(a), vdupq_n_s32(0)) : \
 //          vshlq_n_s32(simde__m128i_to_neon_i32(a), ((imm8) & 31))))
-//#elif defined(SIMDE_WASM_SIMD128_NATIVE)
-//  #define simde_mm_slli_epi32(a, imm8) \
-//    ((imm8 < 32) ? wasm_i32x4_shl(simde__m128i_to_private(a).wasm_v128, imm8) : wasm_i32x4_const(0,0,0,0))
-//#elif defined(SIMDE_POWER_ALTIVEC_P6_NATIVE)
-//  #define simde_mm_slli_epi32(a, imm8) \
-//     (__extension__ ({ \
-//       simde__m128i ret; \
-//       if ((imm8) <= 0) { \
-//         ret = a; \
-//       } else if ((imm8) > 31) { \
-//         ret = simde_mm_setzero_si128(); \
-//       } else { \
-//         ret = simde__m128i_from_altivec_i32( \
-//           vec_sl(simde__m128i_to_altivec_i32(a), \
-//             vec_splats(HEDLEY_STATIC_CAST(unsigned int, (imm8) & 31)))); \
-//       } \
-//       ret; \
-//     }))
+#elif defined(SIMDE_WASM_SIMD128_NATIVE)
+  #define simde_mm_slli_epi32(a, imm8) \
+    ((imm8 < 32) ? wasm_i32x4_shl(simde__m128i_to_private(a).wasm_v128, imm8) : wasm_i32x4_const(0,0,0,0))
+#elif defined(SIMDE_POWER_ALTIVEC_P6_NATIVE)
+  #define simde_mm_slli_epi32(a, imm8) \
+     (__extension__ ({ \
+       simde__m128i ret; \
+       if ((imm8) <= 0) { \
+         ret = a; \
+       } else if ((imm8) > 31) { \
+         ret = simde_mm_setzero_si128(); \
+       } else { \
+         ret = simde__m128i_from_altivec_i32( \
+           vec_sl(simde__m128i_to_altivec_i32(a), \
+             vec_splats(HEDLEY_STATIC_CAST(unsigned int, (imm8) & 31)))); \
+       } \
+       ret; \
+     }))
 #endif
 #if defined(SIMDE_X86_SSE2_ENABLE_NATIVE_ALIASES)
   #define _mm_slli_epi32(a, imm8) simde_mm_slli_epi32(a, imm8)

--- a/thirdparty/simde/x86/sse2.h
+++ b/thirdparty/simde/x86/sse2.h
@@ -5915,10 +5915,9 @@ simde_mm_slli_epi32 (simde__m128i a, const int imm8)
     r_,
     a_ = simde__m128i_to_private(a);
 
-
   #if defined(SIMDE_ARM_NEON_A32V7_NATIVE)
-    const int cnt = (imm8 & ~15) ? 15 : imm8;
-    r_.neon_i16 = vshlq_s32(a_.neon_i32, vdupq_n_s32(HEDLEY_STATIC_CAST(int32_t, cnt)));
+    const int cnt = (imm8 & ~31) ? 31 : imm8;
+    r_.neon_i32 = vshlq_s32(a_.neon_i32, vdupq_n_s32(HEDLEY_STATIC_CAST(int32_t, cnt)));
   #elif defined(SIMDE_VECTOR_SUBSCRIPT_SCALAR)
     r_.i32 = a_.i32 << imm8;
   #else

--- a/thirdparty/simde/x86/sse2.h
+++ b/thirdparty/simde/x86/sse2.h
@@ -2644,8 +2644,8 @@ simde_mm_cvtpd_ps (simde__m128d a) {
 
     #if defined(SIMDE_ARM_NEON_A64V8_NATIVE)
       r_.neon_f32 = vcombine_f32(vcvt_f32_f64(a_.neon_f64), vdup_n_f32(0.0f));
-    #elif defined(SIMDE_POWER_ALTIVEC_P7_NATIVE)
-      r_.altivec_f32 = vec_float2(a_.altivec_f64, vec_splats(0.0));
+//    #elif defined(SIMDE_POWER_ALTIVEC_P7_NATIVE)
+//      r_.altivec_f32 = vec_float2(a_.altivec_f64, vec_splats(0.0));
     #elif defined(SIMDE_WASM_SIMD128_NATIVE)
       r_.wasm_v128 = wasm_f32x4_demote_f64x2_zero(a_.wasm_v128);
     #elif HEDLEY_HAS_BUILTIN(__builtin_shufflevector) && HEDLEY_HAS_BUILTIN(__builtin_convertvector)

--- a/thirdparty/simde/x86/sse4.1.h
+++ b/thirdparty/simde/x86/sse4.1.h
@@ -1237,8 +1237,8 @@ simde_mm_extract_epi32 (simde__m128i a, const int imm8)
 #  define simde_mm_extract_epi32(a, imm8) _mm_extract_epi32(a, imm8)
 #elif defined(SIMDE_ARM_NEON_A32V7_NATIVE)
 #  define simde_mm_extract_epi32(a, imm8) vgetq_lane_s32(simde__m128i_to_private(a).neon_i32, imm8)
-#elif defined(SIMDE_POWER_ALTIVEC_P6_NATIVE)
-#  define simde_mm_extract_epi32(a, imm8) HEDLEY_STATIC_CAST(int32_t, vec_extract(simde__m128i_to_private(a).altivec_i32, imm8))
+//#elif defined(SIMDE_POWER_ALTIVEC_P6_NATIVE)
+//#  define simde_mm_extract_epi32(a, imm8) HEDLEY_STATIC_CAST(int32_t, vec_extract(simde__m128i_to_private(a).altivec_i32, imm8))
 #endif
 #if defined(SIMDE_X86_SSE4_1_ENABLE_NATIVE_ALIASES)
   #undef _mm_extract_epi32

